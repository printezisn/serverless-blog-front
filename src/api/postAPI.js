class PostAPI {
    constructor() {
        this.root = process.env.GATSBY_EDNABLOG_API_ROOT_URL
    }

    async createPost(post) {
        const response = await fetch(this.root, {
            method: "PUT",
            mode: "cors",
            body: JSON.stringify(post),
        })

        return response.json()
    }

    async updatePost(post) {
        const response = await fetch(this.root, {
            method: "POST",
            mode: "cors",
            body: JSON.stringify(post),
        })

        return response.json()
    }

    async getPost(id) {
        const response = await fetch(this.root + encodeURIComponent(id), {
            method: "GET",
            mode: "cors",
        })

        return response.json()
    }

    async deletePost(id) {
        const response = await fetch(this.root + encodeURIComponent(id), {
            method: "DELETE",
            mode: "cors",
        })

        return response.json()
    }
}

export const postAPI = new PostAPI()
