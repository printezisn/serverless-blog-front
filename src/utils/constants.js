export class Events {
    static AUTH_CHANGE = "AuthChange"
}
