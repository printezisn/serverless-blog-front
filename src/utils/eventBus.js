class EventBus {
    constructor() {
        this.events = {}
    }

    register(eventName, listener) {
        this.events[eventName] = this.events[eventName] || []
        this.events[eventName].push(listener)
    }

    unregister(eventName, listener) {
        const listeners = this.events[eventName] || []
        const index = listeners.indexOf(listener)
        if (index >= 0) {
            listeners.splice(index, 1)
        }
    }

    emit(eventName, data) {
        const listeners = this.events[eventName] || []
        listeners.forEach(listener => listener(data))
    }
}

export const eventBus = new EventBus()
