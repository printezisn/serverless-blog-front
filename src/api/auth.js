import { CognitoAuth } from "amazon-cognito-auth-js"
import { Events } from "../utils/constants"
import { eventBus } from "../utils/eventBus"

class Auth {
    constructor() {
        const rootUrl = process.env.GATSBY_EDNABLOG_SITE_ROOT_URL
        const authData = {
            ClientId: "7icdk42lkcj7fevgp2lgss0aio",
            AppWebDomain: "ednablog.auth.ca-central-1.amazoncognito.com",
            TokenScopesArray: ["email", "openid"],
            RedirectUriSignIn: rootUrl,
            RedirectUriSignOut: rootUrl,
            UserPoolId: "ca-central-1_zPiF08WsY",
            AdvancedSecurityDataCollectionFlag: false,
        }

        const self = this
        const auth = new CognitoAuth(authData)
        auth.userhandler = {
            onSuccess: result => {
                self.accessToken = result.getAccessToken().getJwtToken()
                eventBus.emit(Events.AUTH_CHANGE)
            },
            onFailure: () => {},
        }

        this.auth = auth
        this.accessToken = this.auth
            .getSignInUserSession()
            .getAccessToken()
            .getJwtToken()
    }

    logIn() {
        this.auth.getSession()
    }

    logOut() {
        this.auth.signOut()
    }

    init(url) {
        this.auth.parseCognitoWebResponse(url)
    }

    getAccessToken() {
        return this.accessToken
    }

    isLoggedIn() {
        return this.accessToken !== ""
    }
}

export const auth = new Auth()
