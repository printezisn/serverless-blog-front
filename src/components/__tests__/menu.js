import React from "react"
import renderer from "react-test-renderer"

jest.mock("../../api/auth", () => {
    return {
        auth: {
            isLoggedIn: jest.fn(),
            logOut: jest.fn(),
        },
    }
})

import { auth } from "../../api/auth"
import Menu from "../menu"
import { Events } from "../../utils/constants"
import { eventBus } from "../../utils/eventBus"

describe("Menu", () => {
    it("renders correctly when it's not active", () => {
        auth.isLoggedIn.mockReturnValue(false)

        const menu = renderer.create(<Menu />).toJSON()
        expect(menu).toMatchSnapshot()
    })

    it("renders correctly when it's active", () => {
        auth.isLoggedIn.mockReturnValue(false)

        const menu = renderer.create(<Menu />)
        menu.root
            .findByProps({ className: "navbar-burger burger" })
            .props.onClick()
        expect(menu.toJSON()).toMatchSnapshot()
    })

    it("renders correctly when the user is logged in", () => {
        auth.isLoggedIn.mockReturnValueOnce(false).mockReturnValue(true)

        const menu = renderer.create(<Menu />)
        menu.root
            .findByProps({ className: "navbar-burger burger" })
            .props.onClick()
        expect(menu.toJSON()).toMatchSnapshot()

        eventBus.emit(Events.AUTH_CHANGE)

        expect(menu.toJSON()).toMatchSnapshot()
    })

    it("logs the user out when the logout link is clicked", () => {
        auth.isLoggedIn.mockReturnValue(true)

        const menu = renderer.create(<Menu />)
        menu.root
            .findByProps({ className: "navbar-item logout-button" })
            .props.onClick()

        expect(auth.logOut).toHaveBeenCalled()
    })
})
