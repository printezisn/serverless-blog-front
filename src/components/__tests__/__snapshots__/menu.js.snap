// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly when it's active 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu is-active"
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item"
        href="/"
      >
        Home
      </a>
      <a
        className="navbar-item"
        href="/post/read/about"
      >
        About
      </a>
      <a
        className="navbar-item"
        href="/post/read/cookie-policy"
      >
        Cookie Policy
      </a>
      
    </div>
  </div>
</nav>
`;

exports[`Menu renders correctly when it's not active 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu "
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item"
        href="/"
      >
        Home
      </a>
      <a
        className="navbar-item"
        href="/post/read/about"
      >
        About
      </a>
      <a
        className="navbar-item"
        href="/post/read/cookie-policy"
      >
        Cookie Policy
      </a>
      
    </div>
  </div>
</nav>
`;

exports[`Menu renders correctly when the user is logged in 1`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu is-active"
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item"
        href="/"
      >
        Home
      </a>
      <a
        className="navbar-item"
        href="/post/read/about"
      >
        About
      </a>
      <a
        className="navbar-item"
        href="/post/read/cookie-policy"
      >
        Cookie Policy
      </a>
      
    </div>
  </div>
</nav>
`;

exports[`Menu renders correctly when the user is logged in 2`] = `
<nav
  aria-label="main navigation"
  className="navbar"
  role="navigation"
>
  <div
    className="navbar-brand"
  >
    <a
      aria-expanded="false"
      aria-label="menu"
      className="navbar-burger burger"
      onClick={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
      <span
        aria-hidden="true"
      />
    </a>
  </div>
  <div
    className="navbar-menu is-active"
  >
    <div
      className="navbar-start"
    >
      <a
        className="navbar-item"
        href="/"
      >
        Home
      </a>
      <a
        className="navbar-item"
        href="/post/read/about"
      >
        About
      </a>
      <a
        className="navbar-item"
        href="/post/read/cookie-policy"
      >
        Cookie Policy
      </a>
      <a
        className="navbar-item logout-button"
        href="#"
        onClick={[Function]}
      >
        Logout
      </a>
    </div>
  </div>
</nav>
`;
